File: libs/program/src/state.rs
use crate::error::InglError;
use borsh::{BorshDeserialize, BorshSerialize};
use serde::{Deserialize, Serialize};
use solana_program::{
    borsh::try_from_slice_unchecked, native_token::LAMPORTS_PER_SOL,
    program_error::ProgramError, pubkey::Pubkey,
    sysvar::{rent::Rent, Sysvar},
    account_info::AccountInfo,
};

use self::constants::*;
pub mod constants{
    use solana_program::{declare_id, native_token::LAMPORTS_PER_SOL};
    declare_id!("8Tzkyx2vprriFPFF2RYmSykKZ51L3H5ayGmvequZQh2F");


    pub const INGL_TREASURY_ACCOUNT_KEY: &str = "ingl_treasury_account_key";
    pub const INGL_NFT_COLLECTION_KEY: &str = "ingl_nft_collection_newer";
    pub const INGL_MINT_AUTHORITY_KEY: &str = "mint_authority";
    pub const INGL_MINTING_POOL_KEY: &str = "minting_pool";
    pub const COLLECTION_HOLDER_KEY: &str = "collection_holder";
    pub const GLOBAL_GEM_KEY: &str = "global_gem_account";
    pub const GEM_ACCOUNT_CONST: &str = "gem_account";
    pub const FEE_MULTIPLYER: u8 = 10;
    pub const PRICE_TIME_INTERVAL: u8 = 20;
    pub const TREASURY_FEE_MULTIPLYER: u8 = 70;
    pub const MAXIMUM_DELEGATABLE_STAKE: u64 = 5*LAMPORTS_PER_SOL;
    pub const BTC_FEED_PUBLIC_KEY: &str = "9ATrvi6epR5hVYtwNs7BB7VCiYnd4WM7e8MfafWpfiXC";
    pub const SOL_FEED_PUBLIC_KEY: &str = "7LLvRhMs73FqcLkA8jvEE1AM2mYZXTmqfUv8GAEurymx";
    pub const ETH_FEED_PUBLIC_KEY: &str = "6fhxFvPocWapZ5Wa2miDnrX2jYRFKvFqYnX11GGkBo2f";
    pub const BNB_FEED_PUBLIC_KEY: &str = "DR6PqK15tD21MEGSLmDpXwLA7Fw47kwtdZeUMdT7vd7L";
    pub const PD_POOL_KEY: &str = "pd_pool";
    pub const PROPOSAL_KEY: &str ="ingl_proposals";
    pub const COUNCIL_MINT_KEY: &str = "council_mint";
    pub const COUNCIL_MINT_AUTHORITY_KEY: &str = "council_mint_authority";
    pub const AUTHORIZED_WITHDRAWER_KEY: &str = "InglAuthorizedWithdrawer";
    pub const VOTE_ACCOUNT_KEY: &str = "InglVote";
    pub const VOTE_DATA_ACCOUNT_KEY: &str = "InglVoteData";
    pub const STAKE_ACCOUNT_KEY: &str = "staking_account_key";
    pub const TREASURY_ACCOUNT_KEY: &str = "Treasury_account_key";
    pub const T_STAKE_ACCOUNT_KEY: &str = "Temporary_stake_account_key";
    pub const T_WITHDRAW_KEY: &str = "Temporary_withdraw";

    pub const VALIDATOR_ID_SHARE: u64 = 15;
    pub const TREASURY_SHARE: u64 = 13;
    pub const TEAM_SHARE: u64 = 12;
    pub const NFTS_SHARE: u64 = 60;

    pub const VALIDATOR_PROPOSAL_VAL_PHRASE: u32 = 842_542_964;
    pub const GLOBAL_GEMS_VAL_PHRASE: u32 = 546_875_215;
    pub const VALIDATOR_VOTE_VAL_PHRASE: u32 =214_648_321;
    pub const INGL_VOTE_ACCOUNT_DATA_VAL_PHRASE: u32 = 842_154_348;
    pub const GEM_ACCOUNT_VAL_PHRASE: u32 = 516_248_961;
    pub const VOTE_REWARDS_VAL_PHRASE:u32 = 584_625_418;


    pub mod spl_program{
        use solana_program::declare_id;
        declare_id!("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
    }
    pub mod metaplex{
        use solana_program::declare_id;
        declare_id!("metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s");
    }
    pub mod associated_token_program{
        use solana_program::declare_id;

        declare_id!("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL");
    }

    pub mod vote_program{
        solana_program::declare_id!("Vote111111111111111111111111111111111111111");
    }
}



#[derive(BorshDeserialize, BorshSerialize, Copy, Clone, Serialize, Deserialize)]
pub struct VoteInit {
    pub node_pubkey: Pubkey,
    pub authorized_voter: Pubkey,
    pub authorized_withdrawer: Pubkey,
    pub commission: u8,
}


#[derive(BorshSerialize, Copy, Clone, BorshDeserialize)]
pub enum Class {
    Ruby,
    Diamond,
    Sapphire,
    Emerald,
    Serendibite,
    Benitoite,
}

impl Class {
    pub fn get_class_lamports(self) -> u64 {
        LAMPORTS_PER_SOL
            * match self {
                Self::Ruby => 500,
                Self::Diamond => 100,
                Self::Sapphire => 50,
                Self::Emerald => 10,
                Self::Serendibite => 5,
                Self::Benitoite => 1,
            }
    }
}

#[derive(BorshDeserialize, BorshSerialize, Clone)]
pub enum Rarity {
    Common,
    Uncommon,
    Rare,
    Exalted,
    Mythic,
}

#[derive(BorshDeserialize, Clone, BorshSerialize)]
pub struct GlobalGems {
    pub validation_phrase: u32,
    pub counter: u32,
    pub total_raised: u64,
    pub pd_pool_total: u64,
    pub delegated_total: u64,
    pub dealloced_total: u64,
    pub is_proposal_ongoing: bool,
    pub proposal_numeration: u32,
    pub pending_delegation_total: u64,
    pub validator_list : Vec<Pubkey>,//This is not the validator list to display for proposals check instead the ValidatorProposal Struct.
    // pub winners_list: Vec<u32>, // To include next so as to 
}
impl GlobalGems{
    pub fn validate(self) -> Result<Self, ProgramError> {
        if self.validation_phrase != GLOBAL_GEMS_VAL_PHRASE{
            Err(InglError::InvalidValPhrase.utilize(Some("Global Gems")))?
        }
        Ok(self)
    }
    pub fn decode(account: &AccountInfo) -> Result<Self, ProgramError> {
                let a: Self = try_from_slice_unchecked(&account.data.borrow()).unwrap();
                a.validate()
    }
}

#[derive(BorshDeserialize, BorshSerialize)]
pub enum FundsLocation {
    MintingPool,
    PDPool,
    VoteAccount { vote_account_id: Pubkey},
}


#[derive(BorshDeserialize, BorshSerialize)]
pub struct ValidatorVote{
    pub validation_phrase: u32,
    pub proposal_id: Pubkey,
    pub validator_index: u32,
}
impl ValidatorVote{
    pub fn validate(self) -> Result<Self, ProgramError> {
        if self.validation_phrase != VALIDATOR_VOTE_VAL_PHRASE{
            Err(InglError::InvalidValPhrase.utilize(Some("Validator Vote")))?
        }
        Ok(self)
    }
    pub fn decode(account: &AccountInfo) -> Result<Self, ProgramError> {
                let a: Self = try_from_slice_unchecked(&account.data.borrow()).unwrap();
                a.validate()
    }
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct GemAccountV0_0_1 {
    pub struct_id: GemAccountVersions,
    pub validation_phrase: u32,
    pub date_created: u32,
    pub class: Class,
    pub redeemable_date: u32,
    pub numeration: u32,
    pub rarity: Option<Rarity>,
    pub funds_location: FundsLocation,
    pub rarity_seed_time: Option<u32>,
    pub date_allocated: Option<u32>,
    pub last_voted_proposal: Option<Pubkey>,
    pub last_withdrawal_epoch: Option<u64>,
    pub last_delegation_epoch: Option<u64>,
    pub all_withdraws: Vec<u64>,
    pub all_votes: Vec<ValidatorVote>,
}
impl GemAccountV0_0_1 {
    pub fn validate(account_data: Self) -> Result<Self, ProgramError> {
        if account_data.validation_phrase != GEM_ACCOUNT_VAL_PHRASE{
            Err(InglError::InvalidValPhrase.utilize(Some("GemAccountV0_0_1")))?
        }
        Ok(account_data)
    }
}

#[derive(BorshDeserialize, BorshSerialize)]
pub enum GemAccountVersions {
    GemAccountV0_0_1,
    BlanckCase
}
impl GemAccountVersions {
    pub fn decode_unchecked<T: BorshDeserialize>(data: &[u8]) -> Result<T, ProgramError> {
        let version: GemAccountVersions = try_from_slice_unchecked(&data[0..1]).unwrap();
        match version {
            GemAccountVersions::GemAccountV0_0_1 => {
                //Change the code here to represent the conversion to the appropriate version you expect.
                let a: T = try_from_slice_unchecked(data).unwrap();
                Ok(a)
            }
            // GemAccountVersions::AnotherOption => {
            //  Do Something in Here to convert data to the appropriate struct to return
            // }
            _ => {
                Err(InglError::InvalidStructType.utilize(Some("GemAccountVersions deserialize")))}
        }
    }
}


#[derive(BorshDeserialize, BorshSerialize)]
pub struct ValidatorProposal{
    pub validation_phrase: u32,
    pub validator_ids: Vec<Pubkey>,
    pub date_created: u32,
    pub date_finalized: Option<u32>,
    pub votes: Vec<u32>,
    pub winner: Option<Pubkey>
}
impl ValidatorProposal{
    pub fn validate(self) -> Result<Self, ProgramError> {
        if self.validation_phrase != VALIDATOR_PROPOSAL_VAL_PHRASE{
            Err(InglError::InvalidValPhrase.utilize(Some("Validator Proposal")))?
        }
        Ok(self)
    }
    pub fn decode(account: &AccountInfo) -> Result<Self, ProgramError> {
        let a: Self = try_from_slice_unchecked(&account.data.borrow()).unwrap();
        a.validate()
}
}

#[derive(BorshDeserialize, Copy, Clone, BorshSerialize)]
pub struct VoteRewards{
    pub validation_phrase: u32,
    pub epoch_number: u64,
    pub total_reward: u64,
    pub total_stake: u64,
}
impl VoteRewards{
    pub fn validate(self) -> Result<Self, ProgramError> {
        if self.validation_phrase != VOTE_REWARDS_VAL_PHRASE{
            Err(InglError::InvalidValPhrase.utilize(Some("Vote rewards")))?
        }
        Ok(self)
    }
    pub fn decode(account: &AccountInfo) -> Result<Self, ProgramError> {
                let a: Self = try_from_slice_unchecked(&account.data.borrow()).unwrap();
                a.validate()
    }
}
#[derive(BorshDeserialize, BorshSerialize)]
pub struct InglVoteAccountData{
    pub validation_phrase: u32,
    pub total_delegated: u64,
    pub last_withdraw_epoch: u64,
    pub dealloced: u64,
    pub pending_validator_rewards: Option<u64>, // Field is also used to check if there is an ongoing rebalancing or not.
    pub validator_id: Pubkey, //To Reconsider.
    pub pending_delegation_total: u64,
    pub vote_rewards: Vec<VoteRewards>,
}
impl InglVoteAccountData{
    pub fn validate(self) -> Result<Self, ProgramError> {
        if self.validation_phrase != INGL_VOTE_ACCOUNT_DATA_VAL_PHRASE{
            Err(InglError::InvalidValPhrase.utilize(Some("ingl vote account data")))?
        }
        Ok(self)
    }
    pub fn decode(account: &AccountInfo) -> Result<Self, ProgramError> {
                let a: Self = try_from_slice_unchecked(&account.data.borrow()).unwrap();
                a.validate()
    }
}


pub struct VoteState{}
impl VoteState {
    pub fn space()->usize{
        3731
    }
    pub fn min_lamports()->u64{
        Rent::get().unwrap().minimum_balance(3731)
    }
}

File: apps/app/src/services/state.ts
import { field, fixedArray, option, variant, vec } from '@dao-xyz/borsh';
import { PublicKey } from '@solana/web3.js';
import { deserializeUnchecked } from 'borsh';

export enum NftClass {
  Ruby,
  Diamond,
  Sapphire,
  Emerald,
  Serendibite,
  Benitoite,
}

export enum Rarity {
  Common,
  Uncommon,
  Rare,
  Exalted,
  Mythic,
}

export enum Instruction {
  MintNft,
  MintNewCollection,
  Redeem,
  ImprintRarity,
  AllocateSol,
  DeAllocateSol,
  CreateVoteAccount,
  ChangeVoteAccountsValidatorIdentity,
  DelegateSol,
  UnDelegateSol,
  InitRarityImprint,
  RegisterValidatorId,
  CreateValidatorSelectionProposal,
  VoteValidatorProposal,
  FinalizeProposal,
}

class GemAccountVersions {}
@variant(1)
export class GemAccountV0_0_1_Version extends GemAccountVersions {}
@variant(0)
export class BlanckCase_Version extends GemAccountVersions {}

class FundsLocation {}
@variant(0)
export class MintingPoolFundLocation extends FundsLocation {}

@variant(1)
export class PDPoolFundLocation extends FundsLocation {}

@variant(2)
export class VoteAccountFundLocation extends FundsLocation {
  @field({ type: fixedArray('u8', 32) })
  vote_account_id!: number;

  constructor(vote_account_id: number) {
    super();
    this.vote_account_id = vote_account_id;
  }
}

export const BTC_HISTORY_BUFFER_KEY = new PublicKey(
  '9ATrvi6epR5hVYtwNs7BB7VCiYnd4WM7e8MfafWpfiXC'
);
export const SOL_HISTORY_BUFFER_KEY = new PublicKey(
  '7LLvRhMs73FqcLkA8jvEE1AM2mYZXTmqfUv8GAEurymx'
);
export const ETH_HISTORY_BUFFER_KEY = new PublicKey(
  '6fhxFvPocWapZ5Wa2miDnrX2jYRFKvFqYnX11GGkBo2f'
);
export const BNB_HISTORY_BUFFER_KEY = new PublicKey(
  'DR6PqK15tD21MEGSLmDpXwLA7Fw47kwtdZeUMdT7vd7L'
);
export const INGL_PROGRAM_ID = new PublicKey(
  '41z2kpMac1RpH5XnBoKnY6vjmJwdbwc1aHRQszCgbyDv'
);
export const INGL_TREASURY_ACCOUNT_KEY = 'ingl_treasury_account_key';
export const AUTHORIZED_WITHDRAWER_KEY = 'InglAuthorizedWithdrawer';
export const INGL_NFT_COLLECTION_KEY = 'ingl_nft_collection_newer';
export const COUNCIL_MINT_AUTHORITY_KEY = 'council_mint_authority';
export const COLLECTION_HOLDER_KEY = 'collection_holder';
export const INGL_MINT_AUTHORITY_KEY = 'mint_authority';
export const INGL_MINTING_POOL_KEY = 'minting_pool';
export const GLOBAL_GEM_KEY = 'global_gem_account';
export const GEM_ACCOUNT_CONST = 'gem_account';
export const COUNCIL_MINT_KEY = 'council_mint';
export const PROPOSAL_KEY = 'ingl_proposals';
export const VOTE_ACCOUNT_KEY = 'InglVote';
export const TREASURY_FEE_MULTIPLYER = 70;
export const PRICE_TIME_INTERVAL = 20;
export const PD_POOL_KEY = 'pd_pool';
export const FEE_MULTIPLYER = 10;

class Assignable {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(properties: any) {
    Object.keys(properties).map((key) => {
      return (this[key] = properties[key]);
    });
  }
}

export class VoteInit extends Assignable {}
export class GlobalGems extends Assignable {}
// export class VoteRewards extends Assignable {}
// export class ValidatorProposal extends Assignable {}
// export class ProgramVoteAccount extends Assignable {}
// export class InglVoteAccountData extends Assignable {}

class Enum {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(properties: any) {
    if (Object.keys(properties).length !== 1) {
      throw new Error('Enum can only take single value');
    }
    Object.keys(properties).map((key) => {
      this[key] = properties[key];
      this['enum'] = key;
      return this;
    });
  }
}

// export class ClassEnum extends Enum {}
// export class RubyOption extends Assignable {}
// export class DiamondOption extends Assignable {}
// export class SapphireOption extends Assignable {}
// export class EmeraldOption extends Assignable {}
// export class SerendibiteOption extends Assignable {}
// export class BenitoiteOption extends Assignable {}

// export class RarityEnum extends Enum {}
// export class CommonOption extends Assignable {}
// export class UncommonOption extends Assignable {}
// export class RareOption extends Assignable {}
// export class ExaltedOption extends Assignable {}
// export class MythicOption extends Assignable {}

export class MintingPoolOption extends Assignable {}
export class PDPoolOption extends Assignable {}
export class VoteAccountOption extends Assignable {}

// export class GemAccountVersionsEnum extends Enum {}
// export class GemAccountV0_0_1Option extends Assignable {}
// export class BlanckCaseOption extends Assignable {}

const INGL_SCHEMA = new Map([
  [
    VoteInit,
    {
      kind: 'struct',
      fields: [
        ['commission', 'u8'],
        ['node_pubkey', ['u8', 32]],
        ['authority_voter', ['u8', 32]],
        ['authorized_withdrawer', ['u8', 32]],
      ],
    },
  ],
  [
    GlobalGems,
    {
      kind: 'struct',
      fields: [
        ['validation_phrase', 'u32'],
        ['counter', 'u32'],
        ['total_raised', 'u64'],
        ['pd_pool_total', 'u64'],
        ['delegated_total', 'u64'],
        ['dealloced_total', 'u64'],
        ['is_proposal_ongoing', 'u8'],
        ['proposal_numeration', 'u32'],
        ['pending_delegation_total', 'u64'],
        ['validator_list', [['u8', 32]]],
      ],
    },
  ],
  [
    MintingPoolOption,
    {
      kind: 'struct',
      fields: [],
    },
  ],
  [
    PDPoolOption,
    {
      kind: 'struct',
      fields: [],
    },
  ],
  [
    VoteAccountOption,
    {
      kind: 'struct',
      fields: [['id', ['u8', 32]]],
    },
  ],
  // [
  //   ValidatorProposal,
  //   {
  //     kind: 'struct',
  //     fields: [
  //       ['validator_ids', [['u8', 32]]],
  //       ['date_created', 'u32'],
  //       ['date_finalized', { kind: 'option', type: 'u32' }],
  //       ['votes', ['u32']],
  //       ['winner', { kind: 'option', type: ['u8', 32] }],
  //     ],
  //   },
  // ],
]);

export async function decodeInglData<T>(
  classType: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    new (args: any): T;
  },
  buffer: Buffer
) {
  return deserializeUnchecked(INGL_SCHEMA, classType, buffer);
}

export class ValidatorProposal {
  @field({ type: 'u32' })
  public validation_phrase!: number;

  @field({ type: vec(fixedArray('u8', 32)) })
  public validator_ids!: PublicKey[];

  @field({ type: 'u32' })
  public date_created!: number;

  @field({ type: option('u32') })
  public date_finalized!: undefined | number;

  @field({ type: vec('u32') })
  public votes!: number[];

  @field({ type: option(fixedArray('u8', 32)) })
  public winner!: undefined | PublicKey;

  constructor(properties?: {
    validation_phrase: number;
    validator_ids: PublicKey[];
    date_created: number;
    date_finalized: undefined | number;
    votes: number[];
    winner: undefined | PublicKey;
  }) {
    if (properties) {
      this.validation_phrase = properties.validation_phrase;
      this.validator_ids = properties.validator_ids;
      this.date_created = properties.date_created;
      this.date_finalized = properties.date_finalized;
      this.winner = properties.winner;
      this.votes = properties.votes;
    }
  }
}

export class ValidatorVote {
  @field({ type: 'u32' })
  public validation_phrase!: number;

  @field({ type: fixedArray('u8', 32) })
  public proposal_id!: PublicKey;

  @field({ type: 'u32' })
  public validator_index!: number;

  constructor(properties?: {
    validation_phrase: number;
    proposal_id: PublicKey;
    validator_index: number;
  }) {
    if (properties) {
      this.validation_phrase = properties.validation_phrase;
      this.proposal_id = properties.proposal_id;
      this.validator_index = properties.validator_index;
    }
  }
}

export class GemAccountV0_0_1 {
  @field({ type: GemAccountVersions })
  public struct_id!: GemAccountVersions;

  @field({ type: 'u32' })
  public validation_phrase!: number;

  @field({ type: 'u32' })
  public date_created!: number;

  @field({ type: 'u8' })
  public class!: number;

  @field({ type: 'u32' })
  public redeemable_date!: number;

  @field({ type: 'u32' })
  public numeration!: number;

  @field({ type: option('u8') })
  public rarity!: undefined | number;

  @field({ type: FundsLocation })
  public funds_location!: FundsLocation;

  @field({ type: option('u32') })
  public rarity_seed_time!: undefined | number;

  @field({ type: option('u32') })
  public date_allocated!: undefined | number;

  @field({ type: option(fixedArray('u8', 32)) })
  public last_voted_proposal!: undefined | PublicKey;

  @field({ type: option('u64') })
  public last_withdrawal_epoch!: undefined | number;

  @field({ type: option('u64') })
  public last_delegation_epoch!: undefined | number;

  @field({ type: vec('u64') })
  public all_withdraws!: PublicKey[];

  @field({ type: vec(ValidatorVote) })
  public all_votes!: ValidatorVote[];

  constructor(properties?: {
    struct_id: GemAccountVersions;
    validation_phrase: number;
    date_created: number;
    class: number;
    redeemable_date: number;
    numeration: number;
    rarity: undefined | number;
    funds_location: FundsLocation;
    rarity_seed_time: undefined | number;
    date_allocated: undefined | number;
    last_voted_proposal: undefined | PublicKey;
    last_withdrawal_epoch: undefined | number;
    last_delegation_epoch: undefined | number;
    all_withdraws: PublicKey[];
    all_votes: ValidatorVote[];
  }) {
    if (properties) {
      this.struct_id = properties.struct_id;
      this.validation_phrase = properties.validation_phrase;
      this.date_created = properties.date_created;
      this.class = properties.class;
      this.redeemable_date = properties.redeemable_date;
      this.numeration = properties.numeration;
      this.rarity = properties.rarity;
      this.funds_location = properties.funds_location;
      this.rarity_seed_time = properties.rarity_seed_time;
      this.date_allocated = properties.date_allocated;
      this.last_delegation_epoch = properties.last_delegation_epoch;
      this.last_withdrawal_epoch = properties.last_withdrawal_epoch;
      this.last_delegation_epoch = properties.last_delegation_epoch;
      this.all_withdraws = properties.all_withdraws;
      this.all_votes = properties.all_votes;
    }
  }
}
